Mathlib/Tactic/Widget/StringDiagram.lean

```diff
@@ -0,0 +1,331 @@
+  ⟨fun v => v.ident ++ v.indices.foldl (fun s x => s ++ s!"_{x}") ""⟩
+
+/-- The penrose variable assciated with a node. -/
+def Node.toPenroseVar (n : Node) : PenroseVar :=
+  ⟨"E", [n.vPos, n.hPosSrc, n.hPosTar], n.e⟩
+
+/-- The penrose variable assciated with a strand. -/
+def Strand.toPenroseVar (s : Strand) : PenroseVar :=
+  ⟨"f", [s.vPos, s.hPos], s.atom₁.e⟩
+
+/-- Expressions to display as labels in a diagram. -/
+abbrev ExprEmbeds := Array (String × Expr)
+
+/-! ## Widget for general string diagrams -/
+
+open ProofWidgets
+
+/-- The state of a diagram builder. -/
+structure DiagramState where
+  /-- The Penrose substance program.
+  Note that `embeds` are added lazily at the end. -/
+  sub : String := ""
+  /-- Components to display as labels in the diagram,
+  mapped as name ↦ (type, html). -/
+  embeds : HashMap String (String × Html) := .empty
+
+/-- The monad for building a string diagram. -/
+abbrev DiagramBuilderM := StateT DiagramState MonoidalM
+
+open scoped Jsx in
+/-- Build a string diagram from state. -/
+def buildDiagram : DiagramBuilderM (Option Html) := do
+  let st ← get
+  if st.sub == "" && st.embeds.isEmpty then
+    return none
+  let mut sub := "AutoLabel All\n"
+  let mut embedHtmls := #[]
+  for (n, (tp, h)) in st.embeds.toArray do
+    sub := sub ++ s!"{tp} {n}\n"
+    embedHtmls := embedHtmls.push (n, h)
+  sub := sub ++ st.sub
+  return <PenroseDiagram
+    embeds={embedHtmls}
+    dsl={include_str ".."/".."/".."/"widget"/"src"/"penrose"/"monoidal.dsl"}
+    sty={include_str ".."/".."/".."/"widget"/"src"/"penrose"/"monoidal.sty"}
+    sub={sub} />
+
+/-- Add a substance `nm` of Penrose type `tp`,
+labelled by `h` to the substance program. -/
+def addEmbed (nm : String) (tp : String) (h : Html) : DiagramBuilderM Unit := do
+  modify fun st => { st with embeds := st.embeds.insert nm (tp, h )}
+
+open scoped Jsx in
+/-- Add the variable `v` with the type `tp` to the substance program. -/
+def addPenroseVar (tp : String) (v : PenroseVar) : DiagramBuilderM Unit := do
+  let h := <InteractiveCode fmt={← Widget.ppExprTagged v.e} />
+  addEmbed (toString v) tp h
+
+/-- Add instruction `i` to the substance program. -/
+def addInstruction (i : String) : DiagramBuilderM Unit := do
+  modify fun st => { st with sub := st.sub ++ s!"{i}\n" }
+
+/-- Add constructor `tp v := nm (vs)` to the substance program. -/
+def addConstructor (tp : String) (v : PenroseVar) (nm : String) (vs : List PenroseVar) :
+    DiagramBuilderM Unit := do
+  let vs' := ", ".intercalate (vs.map (fun v => toString v))
+  addInstruction s!"{tp} {v} := {nm} ({vs'})"
+
+/-- Run the program in the diagram builder monad. -/
+def DiagramBuilderM.run {α : Type} (x : DiagramBuilderM α) : MonoidalM α :=
+  x.run' {}
+
+open scoped Jsx in
+/-- Construct a string diagram from a Penrose `sub`stance program and expressions `embeds` to
+display as labels in the diagram. -/
+def mkStringDiag (e : Expr) : MonoidalM Html := do
+  DiagramBuilderM.run do
```

Comment:
I wonder, since `DiagramBuilderM` is now being used in more places, maybe it should be upstreamed to ProofWidgets?

Reply:
That sounds reasonable. Should I make such a PR?

Reply:
I already put it there :) However, this PR won't be able to use it until mathlib4 uses the next release of ProofWidgets4.

Reply:
@Vtec234 Now I need a slightly generalized version of `DiagramBuilderM` (see https://github.com/leanprover-community/mathlib4/pull/10581/commits/035db686c6e5cb737315da22f636304e39e26ee0). Would you mind if I make a PR to ProofWidget4 about this change?